File: ArrayHelper/ArrayHelper/ArrayHelper.js
class ArrayHelper {
  static clone(arr) {
    let out = Array.isArray(arr) ? Array() : {};
    for (let key in arr) {
      let value = arr[key];
      if (typeof value.clone === "function") {
        out[key] = value.clone();
      } else {
        out[key] = typeof value === "object" ? ArrayHelper.clone(value) : value;
      }
    }
    return out;
  }

  static equals(arrA, arrB) {
    if (arrA.length !== arrB.length) {
      return false;
    }
    let tmpA = arrA.slice().sort();
    let tmpB = arrB.slice().sort();

    for (var i = 0; i < tmpA.length; i++) {
      if (tmpA[i] !== tmpB[i]) {
        return false;
      }
    }
    return true;
  }

  static print(arr) {
    if (arr.length == 0) {
      return "";
    }
    let s = "(";
    for (let i = 0; i < arr.length; i++) {
      s += arr[i].id ? arr[i].id + ", " : arr[i] + ", ";
    }
    s = s.substring(0, s.length - 2);
    return s + ")";
  }

  static each(arr, callback) {
    for (let i = 0; i < arr.length; i++) {
      callback(arr[i]);
    }
  }

  static get(arr, property, value) {
    for (let i = 0; i < arr.length; i++) {
      if (arr[i][property] == value) {
        return arr[i];
      }
    }
  }

  static contains(arr, options) {
    if (!options.property && !options.func) {
      for (let i = 0; i < arr.length; i++) {
        if (arr[i] == options.value) {
          return true;
        }
      }
    } else if (options.func) {
      for (let i = 0; i < arr.length; i++) {
        if (options.func(arr[i])) {
          return true;
        }
      }
    } else {
      for (let i = 0; i < arr.length; i++) {
        if (arr[i][options.property] == options.value) {
          return true;
        }
      }
    }

    return false;
  }

  static intersection(arrA, arrB) {
    let intersection = new Array();
    for (let i = 0; i < arrA.length; i++) {
      for (let j = 0; j < arrB.length; j++) {
        if (arrA[i] === arrB[j]) {
          intersection.push(arrA[i]);
        }
      }
    }
    return intersection;
  }

  static unique(arr) {
    let contains = {};
    return arr.filter(function (i) {
      return contains[i] !== undefined ? false : (contains[i] = true);
    });
  }

  static count(arr, value) {
    let count = 0;
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] === value) {
        count++;
      }
    }
    return count;
  }

  static toggle(arr, value) {
    let newArr = Array();
    let removed = false;
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] !== value) {
        newArr.push(arr[i]);
      } else {
        removed = true;
      }
    }
    if (!removed) {
      newArr.push(value);
    }
    return newArr;
  }

  static remove(arr, value) {
    let tmp = Array();

    for (let i = 0; i < arr.length; i++) {
      if (arr[i] !== value) {
        tmp.push(arr[i]);
      }
    }

    return tmp;
  }

  static removeUnique(arr, value) {
    let index = arr.indexOf(value);

    if (index > -1) {
      arr.splice(index, 1);
    }
    return arr;
  }

  static removeAll(arrA, arrB) {
    return arrA.filter(function (item) {
      return arrB.indexOf(item) === -1;
    });
  }

  static merge(arrA, arrB) {
    let arr = new Array(arrA.length + arrB.length);
    for (let i = 0; i < arrA.length; i++) {
      arr[i] = arrA[i];
    }
    for (let i = 0; i < arrB.length; i++) {
      arr[arrA.length + i] = arrB[i];
    }
    return arr;
  }

  static containsAll(arrA, arrB) {
    let containing = 0;
    for (let i = 0; i < arrA.length; i++) {
      for (let j = 0; j < arrB.length; j++) {
        if (arrA[i] === arrB[j]) {
          containing++;
        }
      }
    }
    return containing === arrB.length;
  }

  static sortByAtomicNumberDesc(arr) {
    let map = arr.map(function (e, i) {
      return { index: i, value: e.atomicNumber.split(".").map(Number) };
    });
    map.sort(function (a, b) {
      let min = Math.min(b.value.length, a.value.length);
      let i = 0;

      while (i < min && b.value[i] === a.value[i]) {
        i++;
      }
      return i === min
        ? b.value.length - a.value.length
        : b.value[i] - a.value[i];
    });
    return map.map(function (e) {
      return arr[e.index];
    });
  }

  static deepCopy(arr) {
    let newArr = Array();
    for (let i = 0; i < arr.length; i++) {
      let item = arr[i];

      if (item instanceof Array) {
        newArr[i] = ArrayHelper.deepCopy(item);
      } else {
        newArr[i] = item;
      }
    }
    return newArr;
  }
}
module.exports = ArrayHelper;

File: src/ArrayHelper.py
from typing import Any, Union, Callable


class ArrayHelper:
    'A static class containing helper functions for array-related tasks.'

    @staticmethod
    def clone(arr: list) -> list:
        'Returns a copy of the object'
        if not hasattr(arr, '__iter__'):
            return {}
        elif isinstance(arr, dict):
            return {key: value for key, value in arr.items()}
        elif isinstance(arr, (list, tuple)):
            return [value for value in arr]
        elif isinstance(arr, str):
            return {str(i): symbol for i, symbol in enumerate(arr)}

    @staticmethod
    def equals(arrA: list, arrB: list) -> bool:
        'Checks 2 objects for equality'
        return len(arrA) == len(arrB) and all(a == b for a, b in zip(sorted(arrA), sorted(arrB)))

    @staticmethod
    def print(arr: list, sep: str = ', ', beg: str = '(', end: str = ')') -> str:
        'Returns the formatted string of the object'
        if len(arr) == 0:
            return ''
        elements = []
        for item in arr:
            if isinstance(item, str):
                elements.append(item)
            elif hasattr(item, '__iter__'):
                elements.append(ArrayHelper.print(
                    item, sep=',', beg='', end=''))
            else:
                elements.append(str(item.id) if hasattr(
                    item, "id") else str(item))
        return f'{beg}{sep.join(elements)}{end}'

    @staticmethod
    def each(arr: list, callback: Callable) -> None:
        [callback(item) for item in arr]

    @staticmethod
    def get(arr: list, property: Union[str, int], value: Union[str, int]) -> Union[str, int]:
        'Return the array element from an array containing objects, where a property of the object is set to a given value.'
        for item in arr:
            if item[property] == value:
                return item

    @staticmethod
    def contains(arr: dict, options: Any) -> bool:
        'Checks whether or not an array contains a given value. the options object passed as a second argument can contain three properties. value: The value to be searched for. property: The property that is to be searched for a given value. func: A function that is used as a callback to return either true or false in order to do a custom comparison'
        if not options.get('property') and not options.get('func'):
            for item in arr:
                if item == options.get('value'):
                    return True
        elif options.get('func'):
            for item in arr:
                if options['func'](item):
                    return True
        else:
            for item in arr:
                if item.get(options['property']) == options['value']:
                    return True
        return False

    @staticmethod
    def intersection(arrA: list, arrB: list) -> list:
        'Returns an array containing the intersection between two arrays. That is, values that are common to both arrays.'
        return [elem for elem in arrA if elem in arrB]

    @staticmethod
    def unique(arr: list) -> list:
        'Returns an array of unique elements contained in an array.'
        return list(set(arr))

    @staticmethod
    def count(arr: list, value: Any) -> int:
        'Count the number of occurences of a value in an array.'
        return arr.count(value)

    @staticmethod
    def toggle(arr: list, value: Any) -> list:
        'Toggles the value of an array. If a value is not contained in an array, the array returned will contain all the values of the original array including the value. If a value is contained in an array, the array returned will contain all the values of the original array excluding the value.'
        if value in arr:
            new_arr = ArrayHelper.remove_unique(arr, value)
        else:
            arr.append(value)
            new_arr = ArrayHelper.clone(arr)
        return new_arr

    @staticmethod
    def remove(arr: list, value: Any) -> list:
        'Remove a value from an array.'
        return [elem for elem in arr if elem != value]

    @staticmethod
    def removeUnique(arr: list, value: Any) -> list:
        'Remove a value from an array with unique values.'
        clone = ArrayHelper.clone(arr)
        if value in clone:
            del clone[clone.index(value)]
        return clone

    @staticmethod
    def removeAll(arrA: list, arrB: list) -> list:
        'Remove all elements contained in one array from another array.'
        return list(filter(lambda item: item not in arrB, arrA))

    @staticmethod
    def merge(arrA: list, arrB: list) -> list:
        'Merges two arrays and returns the result. The first array will be appended to the second array.'
        arr = [None] * (len(arrA) + len(arrB))
        for i, value in enumerate(arrA):
            arr[i] = value
        for i, value in enumerate(arrB):
            arr[len(arrA) + i] = value
        return arr

    @staticmethod
    def contains_all(arrA: list, arrB: list) -> bool:
        'Checks whether or not an array contains all the elements of another array, without regard to the order.'
        return all([i == j for i in arrA for j in arrB]) and len(arrB) == len(arrA)

    @staticmethod
    def sort_by_atomic_number_desc(arr: list) -> list:
        'Sort an array of atomic number information. Where the number is indicated as x, x.y, x.y.z, ...'
        map_list = list()
        for i, e in enumerate(arr):
            atomic_numbers = list(map(int, e['atomicNumber'].split('.')))
            map_list.append({'index': i, 'value': atomic_numbers})
        map_list.sort(key=lambda x: x['value'], reverse=True)
        return [arr[e['index']] for e in map_list]

    @staticmethod
    def deep_copy(arr: list) -> list:
        'Copies a an n-dimensional array.'
        return [ArrayHelper.deep_copy(elem) if hasattr(arr, '__iter__') else elem for elem in arr]
